### ISSUE TITLE
Will sonnet support tensorfow 2.0?

### ISSUE THREAD

**Issue by lqvito** [deepmind/sonnet#117]

**Comment by tomhennigan** [comment]
Absolutely! We've been discussing with many researchers and engineers at DeepMind and Google Brain how to approach this.

The internals of Sonnet are not compatible with TF2 as they stand, for example Sonnet makes use of graph collections, `tf.get_variable`, `tf.make_template` and `tf.variable_scope` which are all going away in TF2.

We've taken the opportunity to design a TF2 first base class for Sonnet, and we've [proposed up-streaming this into core TensorFlow as `tf.Module`](https://github.com/tensorflow/community/pull/56). Please take a look and comment on the RFC!

We're currently working on rebuilding Sonnet against `tf.Module`, making sure more advanced workflows like using custom getters (e.g. [bayes by backprop](https://github.com/deepmind/sonnet/blob/087cb9d97e374f4bf2b36662dc89f38545ea5fba/sonnet/python/custom_getters/bayes_by_backprop.py)) are still possible. We're also making sure Sonnet will work with [TensorFlow distribution strategy](https://github.com/tensorflow/community/pull/25).

These changes will require a new major release of Sonnet (aka. Sonnet 2). It will feel very familiar for anyone who has used Sonnet before and we hope with `tf.Module` the internals will be even simpler and modules can be closer to the math. We're being mindful to make sure the migration is not expensive for users (e.g. as much as possible using the builtin modules will be the same as V1).

For a preview of how this might look please have a look at [the examples on the RFC](https://github.com/tensorflow/community/blob/976bc376830869726ab7b55cde40886b57e55483/rfcs/20190117-tf-module.md#examples).

We'll be up-streaming our progress on Sonnet 2 over the next month or two :smile:

**Comment by londoed** [comment]
Sounds awesome! Is there any way that a Gitter chat room or something similar could be opened for Sonnet? I think it would help users to have a resource like that when making the switch over to Sonnet 2 and also alert the developers to incompatibilities or general bugs.

**Comment by tomhennigan** [comment]
Sorry for the delay getting back to you @londoed! I don't think a chat room is on the cards for the moment, but we do plan to create great docs for Sonnet 2 and please feel free to file bugs on here, we do actively look into everything that is reported!

**Comment by edoardogiacomello** [comment]
I'm trying to implement the VQ EMA block in tensorflow 2.0, since I come from a GAN background and I'd like to learn more about this approach.
It seems quite straightforward, since if I'm not mistaken there will be no need to enforce control dependencies anymore. 
Still, I'm missing some points, like how the encoding could be updated while still being discrete (won't any update, made either with Ema or a gradient, change the value such that they become non-integer values?) or how to deal with batches of data.

**Comment by malcolmreynolds** [comment]
@edoardogiacomello yes, if you write as TF 2.0 with eager execution, you can just write things in the order you want them to happen, no control dependencies will be required.

As for the discrete values - the discreteness in VQ-VAE is not that things get mapped to integer valued codes, but rather that there is a finite/discrete set of codes in your codebook. So you're right that EMA or gradients will provide small adjustments to the values, but that is all working as intended.

As for batching, the TF / Sonnet 1.0 version supports batched data so I think you can do close to a line-by-line translation, I don't think any big changes are necessary. Alternatively, we will be adding a  Sonnet 2.0 version of VQ VAE in the next few months.

**Comment by tomhennigan** [comment]
Just stumbled back on this thread, @malcolmreynolds added `snt.nets.VectorQuantizer` and `snt.nets.VectorQuantizerEMA` to Sonnet 2 in deepmind/sonnet@67f036929d6be76ef142767a5d088ea60211b8fa.

If it's still relevant for you then you can install Sonnet 2 using the instructions at https://github.com/deepmind/sonnet/tree/v2 and give it a try 😄.

**Comment by tomhennigan** [comment]
FYI Sonnet 2 is now available in beta on PyPi with support for TensorFlow 2. You can find more info in the read me :smile:.
